### PR TITLE
GOTH test for renegotiating proposals

### PR DESCRIPTION
**Pull request by jiivan**
fixes: #1423 

```
# Version after fix applied (should pass)
GITHUB_API_TOKEN=<XXX>  poetry run pytest -svx domain/ya-provider/test_provider_multi_activity.py::test_provider_renegotiate_proposal --config-override docker-compose.build-environment.commit-hash=54917c0bc383613afaf5cfc34279035a142fee58

# Version before fix applied (should fail)
GITHUB_API_TOKEN=<XXX>  poetry run pytest -svx domain/ya-provider/test_provider_multi_activity.py::test_provider_renegotiate_proposal --config-override docker-compose.build-environment.commit-hash=224532666626988377305665574fc6ed45f81f85
```

### CODE REVIEW

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py [hunk 1]
```diff
@@ -91,3 +91,189 @@ async def test_provider_multi_activity(
...
+                .property("golem.srv.caps.multi-activity", True)
+                .constraints(
+                    "(&(golem.com.pricing.model=linear)\
+                    (golem.srv.caps.multi-activity=true)\
+                    (golem.runtime.name=wasmtime))"
+                )
+                .build()
+            )
+
+        async def negotiate_begin(requestor, demand, providers):
```
Comment (zakaprov) [inline goth_tests/domain/ya-provider/test_provider_multi_activity.py:136]: My initial thought here: at least some of these internal helper functions can be moved to `goth_tests/helpers/negotiation.py`.

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py [hunk 2]
```diff
@@ -159,3 +159,184 @@ async def test_provider_single_simultaneous_activity(
...
+                    provider.name,
+                )
+                # await provider.wait_for_proposal_accepted()
+
+                new_proposals = []
+                collected_offers = await requestor.api.market.collect_offers(
+                    subscription_id
+                )
+                logger.info("collected offers: %s", collected_offers)
+                assert len(collected_offers) == 2
```
Comment (tworec) [inline goth_tests/domain/ya-provider/test_provider_multi_activity.py:249]: This binds invocation of this fn to the very specific place within the scenario. If invoked before `requestor-1` is finished we will get only 1 proposal.

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py [hunk 3]
```diff
@@ -159,3 +159,184 @@ async def test_provider_single_simultaneous_activity(
...
+
+            subscription_id, demand = await requestor.subscribe_demand(demand)
+
+            proposals = await requestor.wait_for_proposals(
+                subscription_id,
+                providers,
+                lambda p: p.properties.get("golem.runtime.name") == "wasmtime",
+            )
+            logger.info("Collected %s proposals", len(proposals))
+            return subscription_id, proposals
```
Comment (tworec) [inline goth_tests/domain/ya-provider/test_provider_multi_activity.py:215]: we should assert that `proposals` is not empty

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py [hunk 4]
```diff
@@ -159,3 +159,184 @@ async def test_provider_single_simultaneous_activity(
...
+        {"name": "provider-1", "type": "VM-Wasm-Provider", "use-proxy": True},
+    ]
+    config_overrides.append(("nodes", nodes))
+
+    runner, config = _create_runner(common_assets, config_overrides, log_dir)
+
+    async with runner(config.containers):
+        requestor1, requestor2 = runner.get_probes(probe_type=RequestorProbe)
+        providers = runner.get_probes(probe_type=ProviderProbe)
+        assert providers
```
Comment (tworec) [inline goth_tests/domain/ya-provider/test_provider_multi_activity.py:184]: I've added this line as @zakaprov did in above tests, and indeed `providers` was empty list

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py [hunk 5]
```diff
@@ -159,3 +159,184 @@ async def test_provider_single_simultaneous_activity(
...
+                    == "No capacity available. Reached Agreements limit: 1"
+                )
+                new_proposals.append(collected_offers[1].proposal)
+
+                agreement_id = await requestor.create_agreement(new_proposals[0])
+                await requestor.confirm_agreement(agreement_id)
+                await provider.wait_for_agreement_approved()
+                await requestor.wait_for_approval(agreement_id)
+                agreement_providers.append((agreement_id, provider))
+            return agreement_providers
```
Comment (tworec) [inline goth_tests/domain/ya-provider/test_provider_multi_activity.py:261]: this should be not empty. assert is missing

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py [hunk 6]
```diff
@@ -159,3 +159,184 @@ async def test_provider_single_simultaneous_activity(
...
+
+                agreement_id = await requestor.create_agreement(new_proposals[0])
+                await requestor.confirm_agreement(agreement_id)
+                await provider.wait_for_agreement_approved()
+                await requestor.wait_for_approval(agreement_id)
+                agreement_providers.append((agreement_id, provider))
+
+            await requestor.unsubscribe_demand(subscription_id)
+            logger.info("Got %s agreements", len(agreement_providers))
+            return agreement_providers
```
Comment (tworec) [inline goth_tests/domain/ya-provider/test_provider_multi_activity.py:294]: ditto

**Review by tworec — CHANGES_REQUESTED** [review]

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py (hunk 3 above)
Comment (tworec) [reply goth_tests/domain/ya-provider/test_provider_multi_activity.py:215]: Afterwards I've realised that `wait_for_proposals` returns the same number of proposals what number of providers is on input.

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py [hunk 7]
```diff
@@ -159,3 +159,184 @@ async def test_provider_single_simultaneous_activity(
...
+            for agreement_id, provider in agreement_providers:
+                logger.info(
+                    "%s Running activity on %s. agreement_id: %s",
+                    requestor.name,
+                    provider.name,
+                    agreement_id,
+                )
+                activity_id = await requestor.create_activity(agreement_id)
+                await provider.wait_for_exeunit_started()
+                if second_requestor is not None:
```
Comment (tworec) [inline goth_tests/domain/ya-provider/test_provider_multi_activity.py:307]: not used

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py [hunk 8]
```diff
@@ -91,3 +91,189 @@ async def test_provider_multi_activity(
...
+                counter_proposal_id = await requestor.counter_proposal(
+                    subscription_id, demand, proposal
+                )
+                counter_providers.append((counter_proposal_id, provider))
+            return counter_providers
+
+        async def renegotiate(requestor, counter_providers, subscription_id):
+            logger.info("%s: renegotiate()", requestor.name)
+            agreement_providers = []
+            for counter_proposal_id, provider in counter_providers:
```
[the remark this answers is not in the document]
Comment (jiivan) [reply goth_tests/domain/ya-provider/test_provider_multi_activity.py:236]: The logic was flawed (it worked by accident). It was a leftover of a code that would work with n-providers.

Optimized it for one provider.

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py (hunk 2 above)
Comment (jiivan) [reply goth_tests/domain/ya-provider/test_provider_multi_activity.py:249]: Yes, the reneegotiate() should be called after first requestor finished.

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py (hunk 3 above)
Comment (jiivan) [reply goth_tests/domain/ya-provider/test_provider_multi_activity.py:215]: Yes, the length should be equal to number of providers. Altough I don't think it's the focus of this test I'll add an assert.

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py (hunk 5 above)
Comment (jiivan) [reply goth_tests/domain/ya-provider/test_provider_multi_activity.py:261]: Removed the iteration so list length is always 1. Left the list format to comply with run() which expects a list.

File: goth_tests/domain/ya-provider/test_provider_multi_activity.py (hunk 1 above)
Comment (jiivan) [reply goth_tests/domain/ya-provider/test_provider_multi_activity.py:199]: I thought about it, but I think that this helpers are specific to renegotiation test. The commonly used helper is `goth_tests.helpers.negotiation.negotiate_agreements()`. I can't think of a common use for those custom helpers.

**Review by tworec — APPROVED** [review]

**Comment by zakaprov** [comment]
@tworec I didn't get a chance to review these changes before merging. @jiivan How does this relate to https://github.com/golemfactory/yapapi/pull/551 ?

**Comment by jiivan** [comment]
> @tworec I didn't get a chance to review these changes before merging. @jiivan How does this relate to [golemfactory/yapapi#551](https://github.com/golemfactory/yapapi/pull/551) ?

:( This part tests a scenario where provider rejects proposal and then renegotiates. In yapapi, requestor rejects proposal and then renegotiates.

**Pull request merged**